sklearn/linear_model/_logistic.py

```diff
@@ -1230,20 +1256,28 @@ def fit(self, X, y, sample_weight=None):
         if self.penalty == "elasticnet" and self.l1_ratio is None:
             raise ValueError("l1_ratio must be specified when penalty is elasticnet.")
 
+        xp, is_array_api, device_ = get_namespace_and_device(X)
+        xp_y, _ = get_namespace(y)
+
         if self.penalty is None:
             if self.C != 1.0:  # default values
                 warnings.warn(
                     "Setting penalty=None will ignore the C and l1_ratio parameters"
                 )
                 # Note that check for l1_ratio is done right above
-            C_ = np.inf
+            C_ = xp.inf
             penalty = "l2"
         else:
             C_ = self.C
             penalty = self.penalty
 
+        orig_X_dtype = (
+            X.dtype
+            if hasattr(X, "dtype") and xp.isdtype(X.dtype, "real floating")
+            else xp.float64
+        )
         if solver == "lbfgs":
-            _dtype = np.float64
+            _dtype = _max_precision_float_dtype(xp, device=device_)
```

Comment:
```suggestion
            # TODO: refactor the LBFGS solver to allow for low-precision per-sample
            # gradient computation (to avoid X-size allocations and do the the
            # compute intensive dot products in float32) while keeping float64
            # gradient accumulation (when possible).
            _dtype = _max_precision_float_dtype(xp, device=device_)
```